### PR TITLE
do not precompile packages that have opt out to precompilation

### PR DESCRIPTION
**Pull request by KristofferC**
Fixes https://github.com/JuliaLang/Pkg.jl/issues/722

### CODE REVIEW

File: src/API.jl [hunk 1]
```diff
@@ -487,10 +487,9 @@ function precompile(ctx::Context)
     for pkg in pkgids
         paths = Base.find_all_in_cache_path(pkg)
         sourcepath = Base.locate_package(pkg)
-        if sourcepath == nothing
-            # XXX: this isn't supposed to be fatal
-            pkgerror("couldn't find path to $(pkg.name) when trying to precompilie project")
-        end
+        sourcepath == nothing && continue
+        # Heuristic for when precompilation is disabled
+        occursin("__precompile__(false)", read(sourcepath, String)) && continue
```
Comment (StefanKarpinski) [inline src/API.jl]: Could do a regex match which allows spaces, i.e. `r"\b__precompile__\(\s*false\s*\)"`.

**Comment by NHDaly** [comment]
:c this fix doesn't seem to have changed anything for me in julia `v1.0.1`.. Is it supposed to have made it into that release? I'm sorry I wasn't able to get the prelease working, so I could've tested this before you released it. :/

Should i be seeing this in 1.0.1? I'm not entirely sure what the "backport pending 1.0" label means exactly. Thanks!

**Comment by KristofferC** [comment]
You might need to clear out `.julia/compiled` for this to take effect.

**Comment by NHDaly** [comment]
:( That doesn't seem to have helped. :( Sorry!

Specifically, I'm seeing `StatPlots` precompile every time, the same as the original issue linked above:
https://github.com/JuliaLang/Pkg.jl/issues/722

```
(19:54:10) $ rm -rf ~/.julia/compiled/v1.0/*
zsh: sure you want to delete more than 100 files in /Users/daly/.julia/compiled/v1.0 [yn]? y

(19:54:13) $ julia1
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.0.1 (2018-09-29)
 _/ |\__'_|_|_|\__'_|  |  Official https://julialang.org/ release
|__/                   |

julia1.0> using StatPlots
[ Info: Precompiling StatPlots [60ddc479-9b66-56df-82fc-76a74619b69c]
[ Info: Precompiling Reexport [189a3867-3050-52da-a836-e630ba90ab69]
[ Info: Precompiling Plots [91a5bcdd-55d7-5caf-9e0b-520d859cae80]
[ Info: Precompiling GR [28b8d3ca-fb5f-59d9-8090-bfdbd6d07a71]
[ Info: Precompiling Distributions [31c24e10-a181-5473-b8eb-7969acd0382f]
[ Info: Precompiling IterableTables [1c8ee90f-4401-5389-894e-7a04a3dc0f4d]
[ Info: Precompiling Widgets [cc8bc4a8-27d6-5769-a93b-9d913e69aa62]
[ Info: Precompiling Clustering [aaaa29a8-35af-508c-8bc3-b662a17a0fe5]
[ Info: Precompiling KernelDensity [5ab0869b-81aa-558d-bb23-cbf5423bbe9b]

julia1.0>

(19:54:13) $ julia1
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.0.1 (2018-09-29)
 _/ |\__'_|_|_|\__'_|  |  Official https://julialang.org/ release
|__/                   |

julia1.0> using StatPlots
[ Info: Precompiling StatPlots [60ddc479-9b66-56df-82fc-76a74619b69c]

julia1.0>
```

**Comment by KristofferC** [comment]
Since I am not seeing a single Pkg command there, I assume this is a Julia bug?

The `precompile` statement in Pkg does not precompile StatPlots which is what this issue is about.

**Comment by NHDaly** [comment]
> The `precompile` statement in Pkg does not precompile StatPlots which is what this issue is about.

Oh yes, you're right. Neat, I didn't even know about that command!

> Since I am not seeing a single Pkg command there, I assume this is a Julia bug?

Ah, maybe so. It seems that this actually doesn't fix #722, then! My problem is the same as #722, which also didn't reference any `Pkg` commands. It looks like both systems were showing the same broken behavior then! :) I'll move the discussion back over to that Issue then.

Thanks Kristoffer! :)

**Comment by KristofferC** [comment]
It seems best to move this to Base if it precompiles packages it shouldn't.

**Pull request merged**